c++/src/kj/compat/http.c++

```diff
@@ -2703,12 +2762,207 @@ private:
+      switch (mode) {
+        case Mode::COMPRESS: {
+          KJ_ASSERT(deflateReset(&ctx) == Z_OK, "deflateReset() failed.");
+          break;
+        }
+        case Mode::DECOMPRESS: {
+          KJ_ASSERT(inflateReset(&ctx) == Z_OK, "inflateReset failed.");
+          break;
+        }
+      }
+
+    }
+
+  private:
+    Result pumpOnce() {
+      // Prepares Zlib's internal state for a call to deflate/inflate, then calls the relevant
+      // function to process the input buffer. It is assumed that the caller has already set up
+      // Zlib's input buffer.
+      //
+      // Since calls to deflate/inflate will process data until the input is empty, or until the
+      // output is full, multiple calls to `pumpOnce()` may be required to process the entire
+      // message. We're done processing once either `result` is `Z_STREAM_END`, or we get
+      // `Z_BUF_ERROR` and did not write any more output.
+      size_t bufSize = 4096;
+      Array<kj::byte> buffer = kj::heapArray<kj::byte>(bufSize);
+      ctx.next_out = buffer.begin();
+      ctx.avail_out = bufSize;
+
+      int result = Z_OK;
+
+      switch (mode) {
+        case Mode::COMPRESS:
+          result = deflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Compression failed.");
+          break;
+        case Mode::DECOMPRESS:
+          result = inflate(&ctx, Z_SYNC_FLUSH);
+          KJ_REQUIRE(result == Z_OK || result == Z_BUF_ERROR || result == Z_STREAM_END,
+                      "Decompression failed.");
+          break;
+      }
+
+      return Result {
+        .processResult = result,
+        .buffer = kj::mv(buffer),
+        .size = bufSize - ctx.avail_out,
+      };
+    }
+
+    kj::Vector<Result> processLoop(kj::Maybe<size_t> maxSize) {
+      // Since Zlib buffers the writes, we want to continue processing until there's nothing left.
+      kj::Vector<Result> output;
+      size_t totalBytesProcessed = 0;
+      for (;;) {
+        Result result = pumpOnce();
+
+        auto status = result.processResult;
+        auto bytesProcessed = result.size;
+        if (bytesProcessed > 0) {
+          output.add(kj::mv(result));
+          totalBytesProcessed += bytesProcessed;
+          // TODO(now): maxSize is not the true max size passed to `receive()` on its first call.
+          // It's the maxSize of this current `receieve()` call (consider multi-frame messages).
+          KJ_IF_MAYBE(maxMsgSize, maxSize) {
+            // This is only non-null for `receive` calls, so we must be decompressing. We don't want
+            // the decompressed message to OOM us, so let's make sure it's not too big.
+            KJ_REQUIRE(totalBytesProcessed < *maxMsgSize,
+                "Decompressed WebSocket message is too large");
+          }
+        }
+
+        if ((ctx.avail_in == 0 && ctx.avail_out != 0) || status == Z_STREAM_END) {
```

Comment:
My reading of zlib.h is that you don't need `&& ctx.avail_out != 0` here -- it says that when using `Z_SYNC_FLUSH`, zlib will only allow `ctx.avail_in` to be zero if it actually managed to write everything, otherwise it will keep `avail_in` non-zero. Does that seem right?

I think `|| status == Z_STREAM_END` may also be redundant, as my reading of zlib.h is that `Z_STREAM_END` will only be returned if `Z_FINISH` was requested, which we never do.

Reply:
This first one was more so just defensive coding, I feel like that's probably the case, but then I saw:
```
  If deflate returns with avail_out == 0, this function must be called again
  with the same value of the flush parameter and more output space (updated
  avail_out), until the flush is complete (deflate returns with non-zero
  avail_out).  In the case of a Z_FULL_FLUSH or Z_SYNC_FLUSH, make sure that
  avail_out is greater than six to avoid repeated flush markers due to
  avail_out == 0 on return.
```
A (maybe weird) case that came to my mind was if we could flush the entire compressed input into 4096 bytes, but then not have space for the `0x00 0x00 0xFF 0xFF`. Would zlib set `ctx.avail_in` to 4, or 0? (I suspect it's _probably_ going to keep it at 4, but since I wasn't totally sure, I put the extra check to adhere to the comment above). I'll re-read the header file to try to convince myself of one or the other.

As for the second thing, if we exclude `|| status == Z_STREAM_END`, then [this test](https://github.com/capnproto/capnproto/pull/1505/files#diff-bad743492083c5cd840b54442f4c32da3375e01f43c2fa021c927a4588f5a47eR3030) will hang (we get `Z_STREAM_END` but never terminate the infinite loop). Apparently, implementations can send payloads [that set](https://datatracker.ietf.org/doc/html/rfc7692#section-7.2.3.4) the `BFINAL` bit of the deflate block header, which seems to override the flush type.

Reply:
Ugh, that comment about `avail_out == 0` seems to directly contradict the earlier comment saying "(In particular avail_in is zero after the call if enough output space has been provided before the call.)" Now I don't know what to believe. Maybe zlib guarantees that `avail_in` and `avail_out` are never both left as zero? Need tests for this, but it's tricky since you need to come up with test input that leads to the compressed message ending exactly at a buffer boundary, as well as another test input where it ends one byte past a buffer boundary.

Regarding `Z_STREAM_END`, I guess I was thinking specifically on the sending (compressing) side, you will never see it. On the receiving (decompressing) side, I suppose we need to handle it, but it sounds like it should be considered an error? Can we assert-fail?

Reply:
Here's an interesting result. I've generated a random sequence of 8192 lowercase characters. This is what happens when sending and receiving from a KJ websocket:

### SEND
```
[sendImpl] about to process message, input message size: 8192
[processLoop] starting processLoop
[processLoop] avail_in at start: 8192
[processLoop] iteration: 0
   [processLoop] bytes processed this run: 4096
   [processLoop] avail_in: 0
   [processLoop] avail_out: 0
[processLoop] iteration: 1
   [processLoop] bytes processed this run: 1022
   [processLoop] avail_in: 0
   [processLoop] avail_out: 3074
[sendImpl] Done Processing Message, output size: ; innerMessage.size() = 5118
```
### RECEIVE
```
[receive] about to process TEXT
[processLoop] starting processLoop
[processLoop] avail_in at start: 5118
[processLoop] iteration: 0
    [processLoop] bytes processed this run: 4096
    [processLoop] avail_in: 2559
    [processLoop] avail_out: 0
[processLoop] iteration: 1
    [processLoop] bytes processed this run: 4096
    [processLoop] avail_in: 0
    [processLoop] avail_out: 0
[processLoop] iteration: 2
    [processLoop] bytes processed this run: 0
    [processLoop] avail_in: 0
    [processLoop] avail_out: 4096
[receive] done processing TEXT, output size: 8193
```

Note that the output on the `receive` size is larger than the original input by 1 character because we append the null terminator.

This seems to conflict with what the zlib header says:
```
DEFLATE...
  If the parameter flush is set to Z_SYNC_FLUSH, all pending output is
  flushed to the output buffer and the output is aligned on a byte boundary, so
  that the decompressor can get all input data available so far.  (In
  particular avail_in is zero after the call if enough output space has been
  provided before the call.
```

My reading of this (specifically `avail_in` and `avail_out` being 0 in the compression (send) example) is that zlib ends up reading all of `avail_in` into its internal state, but can't flush everything to the output buffer in one go. So as long as we get `avail_out == 0`, we need to flush again, i.e. we should be checking for `avail_out != 0` if we're using `Z_SYNC_FLUSH`.

---
As for `Z_STREAM_END`, I agree we wouldn't expect to see it when sending, and _realistically_ we wouldn't ever expect to see it when receiving (since it doesn't seem likely someone is using an implementation that doesn't allow sending empty uncompressed blocks). My sense is that we should be treating it as any other message, so it worries me a bit that `zlib` throws `Z_STREAM_END` at me when it sees a block with `bfinal` set. I suspect it shouldn't be breaking like this because [the rfc says](https://datatracker.ietf.org/doc/html/rfc7692#section-7.2.3.4):
```
   The last 1 octet (0x00) contains the header
   bits with "BFINAL" set to 0 and "BTYPE" set to 00, and 5 padding bits
   of 0.  This octet is necessary to allow the payload to be
   decompressed in the same manner as messages flushed using DEFLATE
   blocks with "BFINAL" unset.
```
In this case it's referring to the payload `0xf3 0x48 0xcd 0xc9 0xc9 0x07 0x00 0x00` (first 7 octets are the message, last is the octet described above).

Reply:
We should throw an exception and terminate the connection on `BFINAL`, it is an error for the other end to send that.

Reply:
https://datatracker.ietf.org/doc/html/rfc7692#section-7.2.1 suggests it's not an error.
```
- An endpoint MAY use DEFLATE blocks of any type.

- An endpoint MAY use both DEFLATE blocks with the "BFINAL" bit set
  to 0 and DEFLATE blocks with the "BFINAL" bit set to 1.

- When any DEFLATE block with the "BFINAL" bit set to 1 doesn't end
  at a byte boundary, an endpoint MUST add minimal padding bits of 0
  to make it end at a byte boundary.  The next DEFLATE block follows
  the padded data if any.
```

https://datatracker.ietf.org/doc/html/rfc7692#section-7.2.3.4 provides an example where `BFINAL = 1`

So I think we would be deviating from the spec if we threw an exception here.

Reply:
Oh, ok, then we need to do what the spec says to do. Sorry, I assumed the spec said never to send this. Why would the client ever send BFINAL? Would it only happen on the last message of the stream?

Reply:
I'm not really sure to be honest :slightly_frowning_face:, if zlib see's `BFINAL` it ends the stream so it's no longer useful. If we're using `inbound_no_context_takeover` then we just reset the stream for each message, in which case its fine, but if we're using the same stream for the entire connection, then it only makes sense to send this for the final message.

Messages are typically flushed by including empty deflate blocks without compression, but the RFC says implementations that use a compression platform that doesn't support empty deflate blocks without compression can use `BFINAL` instead.

```
   On platforms on which the flush method using an empty DEFLATE block
   with no compression is not available, implementors can choose to
   flush data using DEFLATE blocks with "BFINAL" set to 1.
```

The only thing that's clear to me is we should still be able to decompress the message. It's unclear whether it should end the stream:
```
       0xf3 0x48 0xcd 0xc9 0xc9 0x07 0x00 0x00

   This is the payload of a message containing "Hello" compressed using
   a DEFLATE block with "BFINAL" set to 1.  The first 7 octets
   constitute a DEFLATE block with "BFINAL" set to 1 and "BTYPE" set to
   01 containing "Hello".  The last 1 octet (0x00) contains the header
   bits with "BFINAL" set to 0 and "BTYPE" set to 00, and 5 padding bits
   of 0.  This octet is necessary to allow the payload to be
   decompressed in the same manner as messages flushed using DEFLATE
   blocks with "BFINAL" unset
```

It's really not obvious what the answer is here, I don't think zlib provides a way to override the default behavior of ending the stream on receiving a `BFINAL`. From what I read online, most implementations will not use `BFINAL` and none of the browsers send it.

Reply:
Well if the other side sent BFINAL, doesn't that imply that they reset their compressor? You can't sent more bytes from the same compression context after BFINAL, right? So resetting the decompressor seems like the right answer to me.

Reply:
https://github.com/capnproto/capnproto/pull/1505/commits/050d7dd2a9b15504635c87b15a158258df367494